backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -0,0 +1,258 @@
+        /// Add the specified report to the datasource.
+        /// The 'From' date is set to the 'To' date of the most recent report, if one exists.
+        /// For every SPL project in the system, a new Snapshot is generated using current project data and a 'snapshotOn' time equal to the newly created reports 'To' date.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to add report.</exception>
+        /// <returns></returns>
+        public ProjectReport Add(ProjectReport report)
+        {
+            if (report.To == null)
+            {
+                report.To = DateTime.UtcNow;
+            }
+            var mostRecentReport = this.Context.ProjectReports
+                .Where(r => r.ReportTypeId == ReportTypes.SPL)
+                .OrderByDescending(r => r.To)
+                .FirstOrDefault();
+            report.From = mostRecentReport?.To;
+            this.Context.Add(report);
+
+            this.Context.ProjectSnapshots.AddRange(GenerateSnapshots((DateTime)report.To));
+
+            this.Context.CommitTransaction();
+
+            return report;
+        }
+
+        /// <summary>
+        /// Update the specified project report metadata in the datasource.
+        /// If the project 'To' date has changed, 
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="RowVersionMissingException">Project report rowversion is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project does not exist.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to edit report.</exception>
+        /// <returns></returns>
+        public ProjectReport Update(ProjectReport report)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var originalReport = this.Context.ProjectReports
+                .SingleOrDefault(p => p.Id == report.Id) ?? throw new KeyNotFoundException();
+
+            if (originalReport.To == null) throw new ArgumentNullException();
+            if (originalReport.IsFinal && report.IsFinal) throw new InvalidOperationException($"Unable to update FINAL project reports.");
+            if (originalReport.From == originalReport.To) throw new InvalidOperationException($"Project report start and end dates cannot be the same.");
+
+            this.Context.Entry(originalReport).CurrentValues.SetValues(report);
+
+            //If the report 'To' date has changed regenerate all the snapshots based on that date.
+            if(!originalReport.To.Equals(report.To))
+            {
+                this.Context.ProjectSnapshots.AddRange(GenerateSnapshots((DateTime)report.To));
```

Comment:
If you are generating new snapshots for this moment in time, then you also need to recalculate the variance for each snapshot.  Seems like you could just call the `Refresh(...)` function at this point.  It should be noted that this action will likely result in different report than what the user is viewing presently.  This means that you will need to ensure the returned report contains the latest snapshot and the UI is updated.  This is particular important before they export the results.

Reply:
Hmm... maybe the way I've done things is confusing.

So first of all I don't persist the variance calculations (I'm assuming you are referring to Baseline Integrity) in the DB at all. Baseline integrity is calculated based on the snapshot values when they are fetched from the backend.

As a result, there should never be a time that I need to update Baseline Integrity when updating a ProjectReport.

This does mean that I need to re-fetch the snapshots when the report changes, but since they are decoupled I think I would need to do that in either case.

Reply:
Variance = Baseline Integrity.  Which is the difference between the Net Proceeds from one snapshot to the next.
I'm wondering based on other comments I've added if it would be wise to store the variance in the DB.
This function will add a bunch of new snapshots to the database for this moment in time.

Not to confuse the subject, but I wonder if we're polluting the database with snapshots for projects when those values haven't changed.  Maybe another conversation for another time.

Reply:
Thinking about this more, I actually think adding the variance to the DB is not a good idea unless we create a join table of some kind.

This is my personal definition of a Project Snapshot: **A snapshot of project financial values at a given time.**

Based on the above, putting baseline integrity directly on a snapshot doesn't make sense, because a snapshot intuitively should not directly depend on another snapshot. Even more so in this case, as the only way to figure out where the baseline integrity came from is to look up the matching snapshot in the ProjectReports table which doesn't even have a foreign key linkage.

Something that could make sense is to create another table, such as SnapshotDiff. This could just be join table between two snapshots, or we could actually store the results of calculations in it. I would probably lean towards the first option.

We could then get associate a list of snapshotDiffs to a report which would greatly simplify the frontend logic.

Reply:
Yes it can go either way.  I originally designed it to be as you said, a snapshot for a moment in time.  However when I saw you added the `NotMapped` property it made me think we might have another option.

There are some pros to making it in the DB.
- performance: we don't need to recalculate each time we look at the report
- consistency: the variance should never change for any given report when fetching data
- network: less data transferred on the wire each time
- because we don't have a hard link (foreign key relationship) with the snapshots it means those snapshots could be deleted and then a report would become invalid.

Reply:
A few comments:

**performance: we don't need to recalculate each time we look at the report.**
Not exactly, we'll still need to calculate this each time we look at unsaved report data (when refreshing data or when changing the from date). However, I agree that making the calculation occur during save/update would result in fewer calculations overall.
**consistency: the variance should never change for any given report when fetching data**
I'm not seeing any increase in consistency. The call to get snapshots occurs completely independently of this call as far as the backend is concerned. Either way the frontend needs to call the api's update, wait for a success, and then call the getsnapshots method.
**network: less data transferred on the wire each time**
I think the opposite. The frontend will send/receive the same amount of data to the API with this change. However, the DB will be required to send/receive the result of the variance calculation to the backend which is net new.
**because we don't have a hard link (foreign key relationship) with the snapshots it means those snapshots could be deleted and then a report would become invalid. Having the variances in the DB removes this issue.**
This is true, although I don't think we have a way for users to delete snapshots currently so more of an issue for future changes in the app.

Reply:
Also the delete method orphans all of its snapshots as well. Could update to selectively delete if no other snapshot is using its To: snapshots.

Reply:
There is also no hard enforcement on the From: value actually pointing to anything here. Theoretically the frontend could just send in a random date that doesn't correspond to any other reports To: date. Should we be enforcing that as well?

Reply:
I did think about that and forgot to add it in the requirements.  An invalid `From` should result in an error returned.  Our `From` dates must match an existing snapshot.

Reply:
Deleting a report doesn't technically orphan the snapshots, as they aren't hard linked.  Which is a good thing too, as we would like to keep historical snapshots of projects.  However it does highlight a design issue where a user wouldn't know there was a snapshot for a given date if there isn't a report to inform them.  Probably a problem we can look into later.

Reply:
performance - if it's saved in the DB we only calculate it once for the report.  The user would have to regenerate it to recalc.
consistency - more to do with the fact that there is no hard link from one snapshot to the next, which means historical data may be purged and now the snapshot can't calculate the original variance.
network - the network is within OpenShift too.  Every byte is cpu/power.

Reply:
I think my only actual qualm with storing the variance is intended use. This isn't really a valid concern since to my knowledge we aren't actually using Snapshots anywhere. We can always rename ProjectSnapshot to ProjectReportSnapshot if it becomes obvious that these snapshots will only be used for reporting.

Reply:
Or we create a new table with variance calculations that is linked to the snapshot table.

Reply:
If we create a separate table are we essentially create a single record for each snapshot?  While that would be normalized, it would be terrible for performance.  So I'd either not store it in the DB, or I'd just use the snapshots.  Your call, pros/cons either way.